packages/@uppy/xhr-upload/src/index.js

```diff
@@ -213,231 +279,32 @@ export default class XHRUpload extends BasePlugin {
-
-      const emitError = (error) => {
-        files.forEach((file) => {
-          this.uppy.emit('upload-error', file, error)
-        })
-      }
-
-      const timer = new ProgressTimeout(this.opts.timeout, () => {
-        const error = new Error(this.i18n('uploadStalled', { seconds: Math.ceil(this.opts.timeout / 1000) }))
-        this.uppy.emit('upload-stalled', error, files)
-      })
-
-      xhr.upload.addEventListener('loadstart', () => {
-        this.uppy.log('[XHRUpload] started uploading bundle')
-        timer.progress()
-      })
-
-      xhr.upload.addEventListener('progress', (ev) => {
-        timer.progress()
-
-        if (!ev.lengthComputable) return
-
-        files.forEach((file) => {
-          this.uppy.emit('upload-progress', file, {
-            uploader: this,
-            bytesUploaded: (ev.loaded / ev.total) * file.size,
-            bytesTotal: file.size,
-          })
-        })
-      })
-
-      xhr.addEventListener('load', (ev) => {
-        timer.done()
-
-        if (this.opts.validateStatus(ev.target.status, xhr.responseText, xhr)) {
-          const body = this.opts.getResponseData(xhr.responseText, xhr)
-          const uploadResp = {
-            status: ev.target.status,
-            body,
-          }
-          files.forEach((file) => {
-            this.uppy.emit('upload-success', file, uploadResp)
-          })
-          return resolve()
-        }
-
-        const error = this.opts.getResponseError(xhr.responseText, xhr) || new Error('Upload error')
-        error.request = xhr
-        emitError(error)
-        return reject(error)
-      })
-
-      xhr.addEventListener('error', () => {
-        timer.done()
-
-        const error = this.opts.getResponseError(xhr.responseText, xhr) || new Error('Upload error')
-        emitError(error)
-        return reject(error)
-      })
-
-      this.uppy.on('cancel-all', ({ reason } = {}) => {
-        if (reason !== 'user') return
-        timer.done()
-        xhr.abort()
-      })
-
-      xhr.open(method.toUpperCase(), endpoint, true)
-      // IE10 does not allow setting `withCredentials` and `responseType`
-      // before `open()` is called.
-      xhr.withCredentials = this.opts.withCredentials
-      if (this.opts.responseType !== '') {
-        xhr.responseType = this.opts.responseType
-      }
+    return file
+  }
 
-      Object.keys(this.opts.headers).forEach((header) => {
-        xhr.setRequestHeader(header, this.opts.headers[header])
-      })
+  async #uploadBundle(files) {
+    const { endpoint } = this.opts
+    const optsFromState = this.uppy.getState().xhrUpload ?? {}
+    const uppyFetch = this.#uppyFetch(files)
+    const { signal } = getUppyAbortController(this.uppy)
+    const body = this.createBundledUpload(files, {
+      ...this.opts,
+      ...optsFromState,
+    })
 
-      xhr.send(formData)
+    await this.uppy.queue.add(async () => {
+      return uppyFetch(endpoint, { ...this.opts, body, signal })
```

Comment:
couldn't we just call uppyFetch here? Why do we need it to return a function?
```suggestion
      return this.#uppyFetch(files)(endpoint, { ...this.opts, body, signal })
```